File: kgg-dec/src/qmc2/ekey.cpp
#include "ekey.h"
#include "base64.h"
#include "tc_tea.h"

#include <algorithm>
#include <array>
#include <string>
#include <vector>

const static std::string kEKeyV2Prefix = "UVFNdXNpYyBFbmNWMixLZXk6";
const static std::array<uint8_t, 16> kEKeyV2Key1{
    0x33, 0x38, 0x36, 0x5A, 0x4A, 0x59, 0x21, 0x40, 0x23, 0x2A, 0x24, 0x25, 0x5E, 0x26, 0x29, 0x28,
};
const static std::array<uint8_t, 16> kEKeyV2Key2{
    0x2A, 0x2A, 0x23, 0x21, 0x28, 0x23, 0x24, 0x25, 0x26, 0x5E, 0x61, 0x31, 0x63, 0x5A, 0x2C, 0x54,
};

template <typename T>
std::span<T> ss2span(std::string_view sv) {
    auto* data = reinterpret_cast<const T*>(sv.data());
    return std::span<T>(const_cast<T*>(data), sv.size());
}

template <typename T>
std::string_view span2ss(std::span<T> span) {
    return std::string_view(reinterpret_cast<char*>(span.data()), span.size());
}

std::vector<uint8_t> decrypt_ekey_v1(std::string_view ekey) {
    std::vector<uint8_t> result = b64_decode(reinterpret_cast<const uint8_t*>(ekey.data()), ekey.size());

    uint32_t tea_key[4] = {
        0x69005600 | static_cast<uint32_t>(result[0] << 16) | (result[1]),
        0x46003800 | static_cast<uint32_t>(result[2] << 16) | (result[3]),
        0x2b002000 | static_cast<uint32_t>(result[4] << 16) | (result[5]),
        0x15000b00 | static_cast<uint32_t>(result[6] << 16) | (result[7]),
    };
    auto decrypted = tc_tea_cbc_decrypt(std::span(result).subspan(8), tea_key);
    if (decrypted.empty()) {
        return {};
    }
    result.resize(8);
    result.insert(result.end(), decrypted.begin(), decrypted.end());
    return result;
}

std::vector<uint8_t> decrypt_ekey_v2(std::string_view ekey) {
    std::vector<uint8_t> result;
    result = tc_tea_cbc_decrypt(ss2span<uint8_t>(ekey), kEKeyV2Key1.data());
    result = tc_tea_cbc_decrypt(std::span(result), kEKeyV2Key2.data());
    return decrypt_ekey_v1(span2ss(std::span(result)));
}

std::vector<uint8_t> decrypt_ekey(std::string_view ekey) {
    if (ekey.starts_with(kEKeyV2Prefix)) {
        ekey.remove_prefix(kEKeyV2Prefix.size());
        return decrypt_ekey_v2(ekey);
    }

    return decrypt_ekey_v1(ekey);
}

File: tmp/lib_um_crypto_rust/um_crypto/qmc/src/ekey.rs
use itertools::Itertools;
use lazy_static::lazy_static;
use std::ops::Mul;
use tc_tea::TcTeaError;
use thiserror::Error;
use umc_utils::base64;

/// Base64 encoded prefix: "QQMusic EncV2,Key:"
const EKEY_V2_PREFIX: &[u8; 24] = b"UVFNdXNpYyBFbmNWMixLZXk6";
const EKEY_V2_KEY1: [u8; 16] = [
    0x33, 0x38, 0x36, 0x5A, 0x4A, 0x59, 0x21, 0x40, 0x23, 0x2A, 0x24, 0x25, 0x5E, 0x26, 0x29, 0x28,
];
const EKEY_V2_KEY2: [u8; 16] = [
    0x2A, 0x2A, 0x23, 0x21, 0x28, 0x23, 0x24, 0x25, 0x26, 0x5E, 0x61, 0x31, 0x63, 0x5A, 0x2C, 0x54,
];

lazy_static! {
    static ref EKEY_SIMPLE_KEY: [u8; 8] = make_simple_key::<8>();
}

#[derive(Debug, PartialEq, Error)]
pub enum EKeyDecryptError {
    #[error("EKey is too short for decryption")]
    EKeyTooShort,
    #[error("Error when decrypting ekey v1: {0}")]
    FailDecryptV1(TcTeaError),
    #[error("Error when decrypting ekey v2: {0}")]
    FailDecryptV2(TcTeaError),
    #[error("Failed to decode b64 content: {0}")]
    Base64Decode(#[from] base64::DecodeError),
}

fn make_simple_key<const N: usize>() -> [u8; N] {
    let mut result = [0u8; N];

    for (i, v) in result.iter_mut().enumerate() {
        let i = i as f32;
        let value = 106.0 + i * 0.1;
        let value = value.tan().abs().mul(100.0);
        *v = value as u8;
    }

    result
}

pub fn decrypt_v1(ekey: &[u8]) -> Result<Vec<u8>, EKeyDecryptError> {
    if ekey.len() < 12 {
        Err(EKeyDecryptError::EKeyTooShort)?;
    }

    let ekey = base64::decode(ekey)?;
    let (header, cipher) = ekey.split_at(8);

    // tea_key: interleave a byte from each stream
    let tea_key = EKEY_SIMPLE_KEY
        .iter()
        .zip(header)
        .flat_map(|(&simple_key_part, &header_part)| [simple_key_part, header_part])
        .collect_vec();

    let plaintext = tc_tea::decrypt(cipher, tea_key).map_err(EKeyDecryptError::FailDecryptV1)?;
    Ok([header, &plaintext].concat())
}

pub fn decrypt_v2(ekey: &[u8]) -> Result<Vec<u8>, EKeyDecryptError> {
    let ekey = base64::decode(ekey)?;
    let ekey = tc_tea::decrypt(ekey, EKEY_V2_KEY1).map_err(EKeyDecryptError::FailDecryptV2)?;
    let ekey = tc_tea::decrypt(ekey, EKEY_V2_KEY2).map_err(EKeyDecryptError::FailDecryptV2)?;
    let ekey = ekey.iter().take_while(|&&b| b != 0).copied().collect_vec();

    decrypt_v1(&ekey)
}

pub fn decrypt<T: AsRef<[u8]>>(ekey: T) -> Result<Vec<u8>, EKeyDecryptError> {
    let ekey = ekey.as_ref();
    match ekey.strip_prefix(EKEY_V2_PREFIX) {
        Some(v2_ekey) => decrypt_v2(v2_ekey),
        None => decrypt_v1(ekey),
    }
}
